### ISSUE TITLE
Assignment to out-of-range element of header stack array

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4-spec#805]
This is a low priority issue, as far as I know, as it seems that most P4 implementations restrict programs that access header stacks to do so at index values that are compile time constant values.  All such references can be checked at compile time for whether they are in or out of range.

Given the recently proposed implementation of run-time variable index expressions, both as R-values and L-values, here:

+ https://github.com/p4lang/p4c/pull/2139
+ https://github.com/p4lang/p4c/issues/2007

it made me wonder what the P4_16 specification said about cases where the run-time index was out of range, if it said anything at all.

If there is an expression `my_hdr_stack[j]` where `j` is out of range, the spec already says this in Section 8.16 "Operations on header stacks":

+ Accessing a header stack hs with an index less than 0 or greater than hs.size results in an undefined value.

When the expression `my_hdr_stack[j]` is used as an R-value, I think "undefined value" is reasonably interpreted similarly to attempting to read an uninitialized field in a header or struct, as described in Section 8.21 "Reading uninitialized values and writing fields of invalid headers".

That leaves open what it would mean to use `my_hdr_stack[j]` as an L-value when `j` is out of range, e.g. an assignment with `my_hdr_stack[j]` or `my_hdr_stack[j].field_name` on the left hand side, or used as an `out` or `inout` parameter, or one of the statements `my_hdr_stack[j].setValid()` or `my_hdr_stack[j].setInvalid()` (that might not be an exhaustive list).

One possibility would be to explicitly document that any modifications to such an out-of-range header stack reference as an L-value results in no modification to any defined value in the program, similar to how the current spec says that assignments to fields within an invalid header should be handled (refer to 3rd-to-last and 2nd-to-last paragraphs of Section 8.21).

I am sure there are other possibilities, but none immediately jump to mind that try to limit the "blast radius" of such an invalid header stack element reference.

Background detail: The behavior in the open source behavioral-model implementation right now is that using such an out of range array index either as an L-value or an R-value results in the bmv2 process throwing an exception and crashing [1].  That seems like a reasonable default behavior for behavioral-model, that can aid in finding bugs in a program using this feature.

[1] See comments in this PR: https://github.com/p4lang/behavioral-model/pull/840/files

**Comment by jnfoster** [comment]
Your proposal and Bmv2's behavior both seem sensible to me.

**Comment by jafingerhut** [comment]
Hmm, there may be some subtleties here I did not realize when creating this issue.

Inside of a parser, a reference to `header_stack.next` is explicitly specified to cause the parser to transition to reject if nextIndex is >= size.  That is definitely not the same as a "no op".

The "no op" behavior proposed in the original comment above I was thinking about when this is done inside of a control.  It could also be specified that way for a header stack run-time index that is not `next` inside of a parser, but that would be different than `header_stack.next` specified behavior.  Maybe that is reasonable, since `.next` is kind of special, anyway.

I will create a PR that tries to make this difference explicit, and people can review it to see what they think.

**Comment by jnfoster** [comment]
I think it is reasonable to let `hdr_stack.next` be special. I guess the alternative is to let an out-of-bounds access cause `reject` in parsers too? That also requires an exception.

**Comment by vgurevich** [comment]
As in our previous discussions on the incorrect behaviors, I would recommend this to be as simple as possible, hopefully the equivalent of accessing an invalid header: reading returns an undetermined result, writing has an unspecified effect. As for `pkt.extract()` it can, indeed, be special (it is kinda special anyway since it is allowed to write into an invalid header :) )

Let is not try to go for no-op, etc. **in the spec** -- if any particular implementation wants to do that, it is fine.

**Comment by jnfoster** [comment]
@vgurevich Here is what the spec [actually says](https://p4.org/p4-spec/docs/P4-16-v1.2.0.html#sec-uninitialized-values-and-writing-invalid-headers) about writes to invalid headers.

> If a write is performed to a field in a currently invalid header, and that header is not part of a header_union, that write must not change any state that is currently defined in the system, neither in header fields nor anywhere else. In particular, that invalid header must remain invalid.
> 
> Either of the kinds of writes to fields in a currently invalid header is allowed to modify state whose values are not defined, e.g. the values of fields in headers that are currently invalid. 

In particular, this behavior is somewhat more constrained than "an unspecified effect."

**Comment by vgurevich** [comment]
@jnfoster, I agree, it is more constrained, although as it is written right now it makes certain common optimizations impossible (or out of spec). But that's a separate problem. 

For now, this rule is probably OK, although if I were to nitpick, I'd say that this rule might or might not apply to the writes outside of the stack boundaries. As the rule states "If a write is performed to a field in a **currently** invalid header". But when you write outside the stack boundaries you write into a header which is **permanently** invalid :)

**Comment by jafingerhut** [comment]
The entire reason this issue was created is because the spec currently does not say what the "blast radius" is for writes to elements of a header stack that are outside of the bounds of the header stack size.  Since most (all?) P4 implementations do not support run-time variable header stack indices, they can trivially be checked at compile time, no problem, and the issue does not arise in such implementations at all.

It still will not arise in those implementations, whether we make it "zero blast radius" or "big blast radius" for implementations that support run-time variable indexes to header stacks.

**Comment by jafingerhut** [comment]
Proposed PR here: https://github.com/p4lang/p4-spec/pull/812
